D alex

D-Alex opened the pull request:
I probably using Roby a bit differently and run into some problems. The following commits improve my workflow but I am not sure how they fit into a bigger picture.

doudou commented on lib/roby/actions/models/action.rb:
```diff
@@ -198,7 +198,11 @@ def run(action_interface, arguments = Hash.new)
                     end
-                    result = action_interface.send(name).as_plan
+                    result = action_interface.send(name)
+                    if(!result.class.has_ancestor?(returned_type))
```
Actually, the #as_plan is really needed as e.g. syskit returns an object that's not a task (but object#as_plan turns it into one).

Why don't you check on the result of #as_plan ?

doudou commented on lib/roby/actions/models/interface.rb:
```diff
@@ -91,7 +91,7 @@ def register_action(name, action_model)
                         const_set task_model_name, task_model
                         task_model.permanent_model = self.permanent_model?
-                        action_model.returns(task_model)
+                        action_model.returns(task_model.superclass)
```
This is a very complicated way of saying returns(Roby::Task). Which can also be said by just deleting the line...

The problem that having a default return type like this was trying to fix is to make sure that action return tasks that allow to identify them. I do start to have the same feeling than you and think that it might be just counter-productive in the end.

doudou commented on lib/roby/relations/dependency.rb:
```diff
@@ -921,7 +921,7 @@ def pretty_print(pp) # :nodoc:
                 pp.text "triggered the failure predicate '#{relation[:failure]}': "
             elsif mode == :unreachable_success
-                pp.text "cannot reach the success condition '#{relation[:success]}': "
+                pp.text "child terminated, success condition can no longer be reached '#{relation[:success]}': "
```
"child terminated" is wrong here. The success condition can be anything, and can fail for other reasons than the task being finished. I like the rephrasing to "success condition can no longer be reached" though.

doudou commented:
A note on the commit prefixes: I use only a few:

  core: for all in lib/roby/ and lib/roby/models/
  coordination: for all in lib/roby/coordination/
  interface: for all in lib/roby/interface/
  templates: for the generation templates

Moreover, I prefer that you squash test and fix in the same commit.

D-Alex commented:
Ok, I rebased everything and removed some of the commits which are obviously not going to be merged. The open issues are:

* http://rock-robotics.org/stable/api/tools/roby/tutorial/shell.html is not working because actions by default generate their own return type 
* Deadlock in case someone calls from a Roby::Task instance something which calls require

I can create an issue for this, if you like.

doudou commented:
Also: the tests don't pass

The best way to develop & test on Roby is to have a separate checkout and then do

~~~
bundle install
bundle exec rake test
~~~

doudou commented:
(and Roby has travis integration enabled ... your PR is being checked right into github !)

doudou commented:
I've moved the return type discussion to a separate issue:  https://github.com/rock-core/tools-roby/issues/46
